[RFC] Incsub 1

Clement0 opened the pull request:
[PR2] This is our second step in the live substitution. For now we implement : 
- A substitution in a whole file (:%s/.../...) open a split containing a sum-up of the sub
- Replacement of the pattern in the split
- Displaying the line number
- Highlighting what has been modified/replace

Next step live substitution as user type his command.

Sorry for the messy git history, we will improve it.

Last question for the test, we are not sure which tests are the most relevant for our function and how to implement them. We tried to find a test in functional or unit directory for a command of ex_cmds.c, without success.

[PR3]The substitution now works live; There are still some issues that need to be fixed. It works well for simple cases.
We implemented it in such a way that it could be, in theory, applied to any other function that has a live mode. This live mode would then be called at each user's input.

We started to implement some lua tests but they do not work yet.

Next, we will implement the different mode (in the split and in the main buffer)

DesbyP committed:
[issue 4615] test null_spec.lua

DesbyP committed:
[issue 4615] test null_spec.lua corrigé

Adrey06 committed:
Create Liens_Utiles

Alternative a notre conv

Commit:
[WIP] issue 4794 : split of a new window for substitute

We will use this window to show the preview of the research and add
an interactive mode

Commit:
[WIP] issue 4794 : split of a new window for substitute

We will use this window to show the preview of the research and add
an interactive mode

vim-patch:7.4.1832
https://github.com/vim/vim/releases/tag/v7.4.1832
https://github.com/vim/vim/commit/dc303bce10c60a3314078ea168064552fadf01af

Commit:
[WIP] issue 4794 : split of a new window for substitute

We will use this window to show the preview of the research and add
an interactive mode

Commit:
Merge branch 'story2' of https://github.com/stretto-editor/neovim into story2

Commit:
Live_sub now print one of the buffer line into our new window

Commit:
Live_sub now print one of the buffer line into our new window

Commit:
Add the line number before the printed line but there's an artefact

Sometimes, after the number, there is a little artefact.
I do not know yet where does it come from.

Commit:
Add the line number before the printed line but there's an artefact

Sometimes, after the number, there is a little artefact.
I do not know yet where does it come from.

Commit:
The highlight function I'd like to use but I'm not sure yet of the
implentation

Commit:
Initialisation of the structure for live_sub

DesbyP committed:
[disp results] backup of matched lines in a klist
 not functional

DesbyP committed:
[disp results] backup of matched lines in a klist
 not functional

DesbyP committed:
[disp results] backup of matched lines in a klist
with a new struct matchedline_T
displayed in the buffer with "l.NB_LINE > LINE"

DesbyP committed:
[disp results] free of structures

DesbyP committed:
[disp results] free of structures

DesbyP committed:
[disp results] less allocations, limited to screen width

DesbyP committed:
[disp results] less allocations, limited to screen width

DesbyP committed:
[disp results] less allocation , reallocat if needed

DesbyP committed:
[disp results] less allocation , reallocat if needed

DesbyP committed:
[disp results] clean-up

DesbyP committed:
[disp results] clean-up

DesbyP committed:
[disp results] backing-up index of every match on each line in a list.

In order to be able to highlight replaced text/pattern, we put the
column numbers in a list.

DesbyP committed:
[disp results] backing-up index of every match on each line in a list.

In order to be able to highlight replaced text/pattern, we put the
column numbers in a list.

Commit:
[disp refsult] Add the highlight of the modifications

Commit:
[disp refsult] Add the highlight of the modifications

DesbyP committed:
[disp results] disp with split only if '%' and there are actual results to display

DesbyP committed:
[disp results] disp with split only if '%' and there are actual results to display

DesbyP committed:
[disp results] correction of warning

DesbyP committed:
[disp results] correction of warning

Adrey06 committed:
Coding Style correction in live_sub function

in order to suit the guidelines of neovim

Adrey06 committed:
Coding Style correction in live_sub function

in order to suit the guidelines of neovim

Adrey06 committed:
Merge branch 'story2' of https://github.com/stretto-editor/neovim into story2

Conflicts:
	src/nvim/ex_cmds.c

Commit:
End of Story2

DesbyP commented:
Forgot to handle all the cases like, for example, "3,35s/foo/bar".
this is now FIXED, there is no split if there are no results or if the substitution takes place on a single line (":s")

justinmk commented:
For testing, you will find "screen tests" most useful. Use https://github.com/neovim/neovim/blob/master/test/functional/ui/bufhl_spec.lua as a starting point (copy it to a new file like `live_substitution_spec.lua`).

I see you're making changes to `do_sub`, but this is not ideal. I recommend implementing the input event, CommandCharPre. The idea is to _make the editor extensible_ such that implementing the "live substitution" can be implemented in vimscript. This makes the changes _much_ more useful because they can be used by many, many other applications, instead of _one specific_ application (live substitution).

bfredl commented:
Also note that you don't need to write the screen expects by hand, `screen:snapshot_util()` can be used at any point in a screen test file to generate an expect at that point. Some notes in https://github.com/neovim/neovim/blob/master/test/functional/ui/screen.lua (but we should probably write more helpful instructions in the wiki, like starting from a template like @justinmk noted)

aym7 commented:
@justinmk Thank you for your advices ! 
We understand the advantages of not modifying the C core but the way to implement it in practice is still vague for us.
We use `do_sub` because we need to recuperate the matched lines in a list. Is there a way to gather the matched lines without modifying `do_sub` or rewriting it ?

bfredl commented:
I don't think @justinmk meant it is always wrong to make changes to `do_sub`, just that it should be reusable to other plugins, like saving the matches in a `v:` variable, instead of hardcoding your behaviors. (But I'm not sure it is necessary, it could be enough to use `search()` or `match()` to find the matches, or maybe not, I haven't looked into the details)

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -59,6 +59,20 @@
+                             Integer line,
+                             Integer col_start,
+                             Integer col_end,
+                             Error *err);
```
Where is this used?

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -59,6 +59,20 @@
 #include "nvim/os/input.h"
 #include "nvim/os/time.h"
 
+//#include "nvim/buffer.c"
```
Path would be `nvim/api/buffer.c`

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -67,6 +81,10 @@ typedef struct sign sign_T;
+#include "ex_cmds_defs.h"
+#include "regexp_defs.h"
+#include "pos.h"
+
```
Includes should be above this `#ifdef` block.  Also missing 'nvim/' prefix in file path.

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -5836,3 +5873,166 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+  // Can't do this recursively.  Can't do it when typing a password. 
+  if (cmdwin_type != 0
+      || cmdline_star > 0
+      ) {
```
Condition fits in one line.

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -5836,3 +5873,166 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+  // Append the lines to our buffer
+  kl_iter(matchedline_T, lmatch, current) {
+    matchedline_T mat = (*current)->data;
+    size_t line_size = sizeof(mat.line) + sizeof(long) + 5;
```
The length of the string placed into `str` is not computed right. `sizeof(mat.line)` is the size of the pointer not the string length. Similar for `sizeof(long)`.

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -5836,3 +5873,166 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+
+    // Reallocation if str not long enough
+    if (line_size > curwin->w_frame->fr_width*sizeof(char))
+      assert(xrealloc(str, line_size) != NULL);
```
The `assert()` around `xrealloc()` can be removed.

ZyX-I replied:
The assert not *can* be removed, it *must* be removed. In some build types with glibc (if NDEBUG is defined) the whole expression will be replaced with `(void) (0)`, effectively removing `xrealloc()` call.

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -5836,3 +5873,166 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+                                      curbuf->handle,
+                                      line,                                     // line in curbuf
+                                      (*col)->data + prefix_size + 1,           // beginning of word
+                                      (*col)->data + prefix_size + match_size); // end of word
```
What if there is more than one match per line and the sub has a different length to what is replaced (`:%s/e/XXXX/g`)?

Commit:
The substitution now works live;

There are still some issues that need to be fixed.
It works well for simple cases.
We implemanted it in such a way that it could be, in theory, applied to any
other function that has a live mode. This live mode would then be called at each
user's input.

We started to implemant some lua tests but they do not work yet.

DesbyP committed:
[live modes] declaration of a global variable to define in what live mode we are

DesbyP committed:
[live disp] changing the name of EVENT_COLON to LIVE_MODE

DesbyP committed:
[live disp] better, OK when ENTER is typed, still stuck when ESC or CTRL-C

DesbyP commented:
Thanks @oni-link, we applied your advices in this commit.

Commit:
Fix one bug (one undo had been forgotten)

but we still need to fix the closing of window when we go from
LS_ONE_WD to LS_NO_WD

aym7 committed:
doesn't use live_cmd anymore

Commit:
add comments

Commit:
Removal the hightligting of the results

DesbyP committed:
[live disp] better, OK when ENTER is typed, still stuck when ESC : can't quit

Commit:
Add a test in is_live if cmdline is NULL

Commit:
enable to change sub mode with livesub=i where i is an integer

Commit:
enable to choose the size of split in livesub with set livesubheight=i where i is an integer

aym7 committed:
free the klist

DesbyP committed:
Revert "enable to choose the size of split in livesub with set livesubheight=i where i is an integer"

This reverts commit 2044d2da79b3ea3fa693021aa827e3b246c43149.

DesbyP committed:
Revert "enable to change sub mode with livesub=i where i is an integer"

This reverts commit 6d41b268ed967127c643cce635520aae19fdfc14.

DesbyP committed:
Merge branch 'inprogress' of https://github.com/stretto-editor/neovim into inprogress

DesbyP committed:
[live disp] still better but not yet

DesbyP commented:
[WIP] almost there but still having some issues and bugs.
Don't hesitate to check it out.

fmoralesc commented on src/nvim/ex_docmd.c:
```diff
@@ -9520,3 +9534,28 @@ static void ex_terminal(exarg_T *eap)
+/// to a live command.
+/// At the moment, only substitute has a live command.
+
+int is_live (char_u *cmd_live)
```
why not a `bool`?

fmoralesc commented on src/nvim/normal.c:
```diff
@@ -5195,7 +5198,7 @@ static void nv_dollar(cmdarg_T *cap)
   cap->oap->inclusive = true;
   /* In virtual mode when off the edge of a line and an operator
-   * is pending (whew!) keep the cursor where it is.
+   * is pending (whew!) keep the cursor wdo_shere it is.
```
typo

bfredl commented on src/nvim/ex_cmds.c:
```diff
@@ -1258,7 +1260,7 @@ static void do_filter(
  * When "cmd" is NULL start an interactive shell.
  */
-void 
+void
```
feel free to reformat signatures of functions you change to the new format (entire signature in one line if it fits in 80 characters) but otherwise please not change it.

eric-burel commented on src/nvim/ex_cmds.c:
```diff
@@ -5836,3 +5867,328 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
 }
+
+/// return the size of a long in term of digit
+int width_long(long nb) {
```
Isn't it the `log10()` function you are trying to implement ? Seemes to be a core C function

DesbyP committed:
[disp result] width_long() replaced by log10()+1

DesbyP committed:
[disp result] width_long() replaced by log10()+1

Commit:
Revert "Revert "enable to change sub mode with livesub=i where i is an integer""

This reverts commit d9694a1fa2c551847fb9f9385fbc4994358bb38f.

aym7 committed:
fixed xcalloc bug

Commit:
Enable two modes for livesub

Commit:
Merge branch 'inprogress2' of https://github.com/stretto-editor/neovim into inprogress2

Adrey06 committed:
enum cmd progress

Adrey06 committed:
Merge branch 'inprogress2' of https://github.com/stretto-editor/neovim into inprogress2

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress2' into inprogress2

Commit:
enable to switch between livesub and nolivesub

DesbyP committed:
[live disp] ok when ESC or CTRL_C

Commit:
Merge branch 'inprogress2' of https://github.com/stretto-editor/neovim into inprogress2

Commit:
Merge branch 'inprogress2' of https://github.com/stretto-editor/neovim into inprogress2

Commit:
bug commented

Commit:
bug fixed

Commit:
Merge branch 'inprogress2' of https://github.com/stretto-editor/neovim into inprogress2

DesbyP committed:
[live disp] TOREDO : ok when ESC or CTRL_C

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress2' into inprogress2

DesbyP committed:
warnigns clean-up

Commit:
livesub enable

Commit:
Merge branch 'inprogress2' of https://github.com/stretto-editor/neovim into inprogress2

justinmk commented on src/nvim/options.lua:
```diff
@@ -2766,5 +2766,14 @@ return {
+      redraw={'everything'},
+      varname='p_sub',
+      defaults={if_true={vi=false, vim=false}}
+    },
```
I believe the problem is you added this in the wrong spot. Try putting it in alphanumeric order: after the `listchars` entry.

Clement0 replied:
It works thanks !

jamessan replied:
That requirement should probably be added to the checklist in option.c.

Commit:
solve vimrc option issue

ZyX-I commented on src/nvim/ex_cmds.c:
```diff
@@ -5836,3 +5873,166 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+
+    // Add the line number to the string
+    // TODO : enhance the display of the line number 
+    sprintf(str, "l.%ld > %s", mat.lnum, (char*)mat.line);
```
`sprintf` is not allowed, use `snprintf(str, line_size, …)`. Must have a space before `*`. I am also not sure that this cast is needed at all.

Adrey06 committed:
update screen in live sub

Commit:
add eventslash = 0 when return

DesbyP committed:
[live disp] highlighting of line number segfault, commented
highlighting of pattern restored

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

Conflicts:
	src/nvim/ex_cmds.c

Commit:
add livesub help

Commit:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

DesbyP committed:
[live disp] close split and update screen smarter

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

Adrey06 committed:
redraw of typedef livesub_state

Adrey06 committed:
conflict resolved : here the buffer is close at every calls

DesbyP committed:
[live disp] close split and buffer smarter

Adrey06 committed:
make a substitution when :%s/ enter

Adrey06 committed:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

DesbyP committed:
[live disp] sprintf replaced by snprintf

Commit:
the undo historic is now free of useless "do_sub"s

Commit:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

aym7 committed:
check if buf is still valid before deleting it

Adrey06 committed:
BUG fixed on highlighting different words in the same line

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

DesbyP committed:
[live disp] handling wildcard

Commit:
issue with width of colum solved

Commit:
some functional tests

Commit:
properly quits the live_mode when user erases characters

DesbyP commented:
This is the last day of our project. We still have an issue when aborting the command with ESC or CTRL_C :
we do a normal_enter(true, true) so insert mode and other things are unlocked, when we type command like :w filename it seems to work, but :q! doesn't quit
when typing ENTER there's no problem

Alongside we're focusing on passing tests.

Tips are welcome !

Commit:
bug fixed on macos

Adrey06 committed:
resolving conflict on correcting bug on highlight

DesbyP committed:
[live disp] `cmdwin_type = 0` set in finish_live_cmd() fixed `:q!` issue

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

Commit:
Replaced calloc and free by xcalloc and xfree in compute_line_number

DesbyP committed:
[live disp] rename LIVE_MODE in EVENT_COLON

Commit:
sub do not open a split by default

Commit:
Merge branch 'inprogress3' into story2

Commit:
correction of functional test

Commit:
Correction of the historic bug

Going into the historic and passing through %s/ type commands had
the unpleasant action to undo their doing.

So if you wanted to do the same substitution as before by going
into the history, you just couldn't

Commit:
add a line in order to avoid unwanted undo in live_sub

aym7 committed:
in live mode, do_sub doesn't add search to history each time

Commit:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

Commit:
iLive mode dont display error messages anymore

Commit:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

DesbyP committed:
Fix Warning highest_num_line may be use uninitialized

Commit:
merge with master

Commit:
Bug fix: undo properly done quitting a live_function

Commit:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

DesbyP committed:
Fix bug with flag '/c'

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

Commit:
Merge branch 'inprogress3' into story2

Commit:
correction of coding style

Commit:
Merge branch 'inprogress3' into story2

Commit:
misstyped

Commit:
Merge branch 'inprogress3' into story2

DesbyP committed:
functional tests
nolivesub > old behaviour
livesun > new behaviou, split while typing cmg, closed after, verify that insertion still works after that

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

aym7 committed:
coding style ex_docmd.c

aym7 committed:
coding style ex_getln and normal files

aym7 committed:
coding style ex_cmd, globals, state and undo

aym7 committed:
fixed merge conflicts

aym7 committed:
Merge branch 'inprogress3' into story2

aym7 committed:
coding style ex_docmd, normal, state

Commit:
fix lines 5900 -> endoff

aym7 committed:
coding style ex_cmds.c

aym7 committed:
Merge branch 'inprogress3' into story2

Commit:
Merge branch 'inprogress3' of https://github.com/stretto-editor/neovim into inprogress3

Robinhola commented:
Hello! We are currently working towards making travis validate our build. We have written some documentation for any one who would like to implement a new live function using our work. Here it is:

> How to implement an existing function into a new live function the same way “live_sub()” reimplements “do_sub()”:
>
>### 1°) Write the live version of the old function
>
>The first step is to write your function. You should keep in mind that this function will be executed at each new user input whenever the parsing detects its call. Thus, you shouldn't change the overall functioning of the function you are reimplementing live and think of a few things to apprehend before testing your function. 
>
>Firstly, each command line sent to your function can come from two different kind of user input: 
>	- the user has written a new character 
>	- the user has erased its last written character
>
>You may use static global variables to keep the last state in mind.	
>	
>Secondly, you should let the user bypass it: the outcome should be the same as the classic version but you should aslo add a vimrc option in order to let the user completly disable it (see 4° for vimrc).
>
>### 2°) Make it called instead of the classic version
>
>Go to “ex_cmds.lua”, find the name of the command you whish to make live and change the name of the former function called when the command is parsed by the name of your function.
>
>### 3°) Make it detected as a live function
>
>The function “is_live()” in “ex_docmd.c” parses the command line and returns true if the command (CMD_) found by the parsing has a live implementation. You should add the command (CMD_) you are making into a live command to the list of values for which it returns “true”.
>
>### 4°) Let the user disable it if he wants
>
>The whole procedure to add an option in the vimrc is described at the begining of “option.c”.

fmoralesc commented on runtime/doc/options.txt:
```diff
@@ -3960,6 +3960,12 @@ A jump table for the options with a short description can be found at |Q_op|.
+'Livesub' 'lsu'		boolean	(default off)
+			global
+					If sets, substitution become dynamics. It also opens a
+					split which displays the line of the occurences of the search.
```
Suggestion:
> If set, substitutions (|:s|) are updated live while the user types the command. 
   A split is open which displays the lines where the search matches.

fmoralesc commented on runtime/doc/options.txt:
```diff
@@ -3960,6 +3960,12 @@ A jump table for the options with a short description can be found at |Q_op|.
 	Strings to use in 'list' mode and for the |:list| command.  It is a
 	comma separated list of string settings.
+						*'livesub'* *'lsu'* *'nolivesub'* *'nlsu'*
+'Livesub' 'lsu'		boolean	(default off)
```
'Livesub' -> 'livesub'

fmoralesc commented on src/nvim/ex_cmds.c:
```diff
@@ -2915,21 +2919,23 @@ void do_sub(exarg_T *eap)
+  static int do_all = false;            // do multiple substitutions per line
+  static int do_ask = false;            // ask for confirmation
+  static bool do_count = false;         // count only */
+  // static int do_error = TRUE;         // if false, ignore errors
```
Maybe document why the initial value was changed? If so, remove this line.

DesbyP committed:
Fix tests lua

Commit:
coding style is now ok

DesbyP committed:
Fix bug with ':s'
Fix test lua

DesbyP committed:
Merge remote-tracking branch 'origin/inprogress3' into inprogress3

aym7 committed:
added soutenance file

aym7 committed:
Revert "added soutenance file"

This reverts commit 45da09ce0288148c366bee60c461dab12c362a06.

DesbyP committed:
Merge branch 'inprogress3' into story2

oni-link commented on src/nvim/ex_cmds.c:
```diff
@@ -2939,13 +2945,25 @@ void do_sub(exarg_T *eap)
+  while (*c != '\0') {
+    c++;
+  }
+  last_is_slash = (*(c-1) == '/');
```
Can `eap->arg` point to an empty string?

DesbyP committed:
Fix one memory leak

Commit:
improve documentation for livesub

Commit:
Merge branch 'inprogress3' into story2

Commit:
improve documentation

KillTheMule commented:
Eric told us you'd not be available for cleanup. I've taken it up in https://github.com/neovim/neovim/pull/4915, thanks for all your work!
